packages/asset-emitter/src/asset-emitter.ts

```diff
@@ -0,0 +1,964 @@
+import {
+  type EmitContext,
+  type Model,
+  type Namespace,
+  type Program,
+  type Type,
+  compilerAssert,
+  getTypeName,
+  isTemplateDeclaration,
+  joinPaths,
+} from "@typespec/compiler";
+import { CustomKeyMap } from "./custom-key-map.js";
+import { Placeholder } from "./placeholder.js";
+import { resolveDeclarationReferenceScope } from "./ref-scope.js";
+import { ReferenceCycle } from "./reference-cycle.js";
+import { TypeEmitter } from "./type-emitter.js";
+import {
+  type AssetEmitter,
+  CircularEmit,
+  type ContextState,
+  Declaration,
+  type EmitEntity,
+  type EmitTypeReferenceOptions,
+  EmitterResult,
+  type EmitterState,
+  type LexicalTypeStackEntry,
+  type NamespaceScope,
+  NoEmit,
+  RawCode,
+  type Scope,
+  type SourceFile,
+  type SourceFileScope,
+  type TypeEmitterMethod,
+  type TypeSpecDeclaration,
+} from "./types.js";
+
+/**
+ * Represent an entry in the reference chain.
+ */
+interface ReferenceChainEntry {
+  method: string;
+  type: Type;
+  context: ContextState;
+}
+
+export function createAssetEmitter<T, TOptions extends object>(
+  program: Program,
+  TypeEmitterClass: typeof TypeEmitter<T, TOptions>,
+  emitContext: EmitContext<TOptions>,
+): AssetEmitter<T, TOptions> {
+  const sourceFiles: SourceFile<T>[] = [];
+
+  const options = {
+    noEmit: program.compilerOptions.noEmit ?? false,
```

Comment:
Should this be using dryRun now?
```suggestion
    noEmit: program.compilerOptions.dryRun ?? false,
```